collector/arp_linux.go

```diff
@@ -73,40 +94,54 @@ func getTotalArpEntries(deviceEntries []procfs.ARPEntry) map[string]uint32 {
 	return entries
 }
 
-func getTotalArpEntriesRTNL() (map[string]uint32, error) {
+func getArpEntriesRTNL() (map[string]uint32, map[string]map[string]int, error) {
 	conn, err := rtnl.Dial(nil)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 	defer conn.Close()
 
 	// Neighbors will also contain IPv6 neighbors, but since this is purely an ARP collector,
 	// restrict to AF_INET.
 	neighbors, err := conn.Neighbours(nil, unix.AF_INET)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
 	// Map of interface name to ARP neighbor count.
 	entries := make(map[string]uint32)
+	// Map of map[InterfaceName]map[StateName]int
+	states := make(map[string]map[string]int)
 
 	for _, n := range neighbors {
 		// Skip entries which have state NUD_NOARP to conform to output of /proc/net/arp.
-		if n.State&unix.NUD_NOARP == 0 {
-			entries[n.Interface.Name]++
+		if n.State&unix.NUD_NOARP != unix.NUD_NOARP {
```

Comment:
Maybe I can say?

```suggestion
		if n.State != unix.NUD_NOARP {
```

Reply:
No, `n.State` is a bit mask. The bitwise AND is necessary to isolate the `NUD_NOARP` flag that we wish to test for. Other bits may be set in `n.State`, so you cannot just do a simple comparison as you suggest.
c.f. `man 7 rtnetlink`:
```
              ndm_state is a bit mask of the following states:
              NUD_INCOMPLETE   a currently resolving cache entry
              NUD_REACHABLE    a confirmed working cache entry
              NUD_STALE        an expired cache entry
              NUD_DELAY        an entry waiting for a timer
              NUD_PROBE        a cache entry that is currently reprobed
              NUD_FAILED       an invalid cache entry
              NUD_NOARP        a device with no destination cache
              NUD_PERMANENT    a static entry
```

Reply:
Thanks for explanation, makes sense. Then current implementation looks ok. I changed `n.State&unix.NUD_NOARP == 0` to have early return, can revert that if you wish.

Reply:
Fixed the bug after i changed here, btw.